### ISSUE TITLE
About partial template and local overwrite

### ISSUE THREAD

**Issue by cveneziani** [hanami/view#139]
I've encountered two behaviours that I didn't expect regarding locals and partial templates.

## Exposure overwrite

When rendering a partial template with a local, it overwrites my exposure that has the same name as the partial local.

Here is an example:

```ruby
# apps/web/controllers/home/show.rb
module Web::Controllers::Home
  class Show
    include Web::Action

    expose :name

    def call(params)
      @name = 'Hanami'
    end
  end
end
```

```erb
<!-- apps/web/templates/home/show.html.erb -->
<h1>Home page</h1>

<%= name %> <!-- Hanami -->
<%= render partial: 'some_partial', locals: { name: 'Partial' } %>
<%= name %> <!-- Partial -->
```

Then it means that I can't rely anymore on the exposed `name`. 

## Partial local available in calling template

```ruby
# apps/web/controllers/home/show.rb
module Web::Controllers::Home
  class Show
    include Web::Action

    # nothing exposed

    def call(params)
    end
  end
end
```

```erb
<!-- apps/web/templates/home/show.html.erb -->
<h1>Home page</h1>

<%= render partial: 'some_partial', locals: { name: 'Partial' } %>
<%= name %> <!-- Partial -->
```

I didn't expect to have access from my template to a local defined for a specific partial template.

Are these two use cases intended to work this way?

**Comment by AlfonsoUceda** [comment]
@cveneziani thanks for spotting this problem, I'll take care of this.

**Comment by jodosha** [comment]
@cveneziani Thanks for the report.

---

The first case ("Exposure overwrite") works as intended: if you specify an inline override it takes the precedence over the broader scope value. Conceptually it's like:

```ruby
foo = "hello"

def call
  foo = "world"
  puts foo
end

call
# => "world"
```

---

For the second case ("Partial local available in calling template") I think that is a problem that should be fixed.

**Comment by cveneziani** [comment]
I agree that in the context of the partial the inline override takes precedence over the broader scope value. My problem is later on, my initial template gets the overwritten value.

If I take back your example, I expect to have this behavior:

```ruby
foo = "hello"

def call
  foo = "world"
  puts foo
end

call
# => "world"
foo 
# => "hello"
```

But in fact, I get that: 

```ruby
foo = "hello"

def call
  foo = "world"
  puts foo
end

call
# => "world"
foo 
# => "world"
```

Is that the intended behavior?

**Comment by jodosha** [comment]
> Is that the intended behavior?

Nope. 😓 

---

In my understanding this is only one problem not two, is it correct?

**Comment by cveneziani** [comment]
Kind of.

The first problem is that the exposure is overwritten.

The second problem is that the action is not exposing anything and suddenly the local from the partial is now available within the initial template.

If I take back your example, the second problem is this one:

```ruby
# no variable defined here

def call
  foo = "world"
end

foo
# => "world"
# instead of undefined local variable or method `foo'
```

**Comment by AlfonsoUceda** [comment]
@cveneziani thanks for the detailed answer, let me fix it ;)

**Comment by AlfonsoUceda** [comment]
Closed by #140
